### PR TITLE
[FLINK-17292][docs] Translate Fault Tolerance training lesson to Chinese

### PR DESCRIPTION
**Pull request by RocMarshal**
## What is the purpose of the change

*(Translate Fault Tolerance training lesson to Chinese)*

## Brief change log

  - *Translate Fault Tolerance training lesson to Chinese*

## Verifying this change

This change is "Translate Fault Tolerance training lesson to Chinese" without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by RocMarshal** [comment]
Hi, @wuchong 

I have finished the first draft of this translation(https://github.com/apache/flink/pull/12727). Could you please check it for me if you have free time?

Thank you.

Best Roc.

**Comment by wuchong** [comment]
@klion26  could you help to review this?

File: docs/learn-flink/fault_tolerance.zh.md [hunk 1]
```diff
@@ -29,180 +29,156 @@ under the License.
...
-- Nothing is lost or duplicated (_exactly once_)
+- Flink 不会努力从故障中恢复（_at most once_）
+- 没有任何丢失，但是你可能会得到重复冗余的结果（_at least once_）
+- 没有丢失或冗余重复（_exactly once_）
 
-Given that Flink recovers from faults by rewinding and replaying the source data streams, when the
-ideal situation is described as **exactly once** this does *not* mean that every event will be
-processed exactly once. Instead, it means that _every event will affect the state being managed by
-Flink exactly once_. 
+假设 Flink 通过倒回和重新发送 source 数据流从故障中恢复，当理想情况被描述为 **精确一次（exactly once）** 时，这并*不*意味着每个事件都将被精确一次处理。
```
Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:147]: `倒回` 改成 `回退` 会好一些吗？
这个地方为什么要加”假设“呢？

File: docs/learn-flink/fault_tolerance.zh.md [hunk 2]
```diff
@@ -29,180 +29,156 @@ under the License.
...
+相反，这意味着 _每一个事件都会影响 Flink 管理的状态精确一次_。
```
Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:148]: 建议这行和上面的进行合并，否则这两行的文字中间会多一个空格

File: docs/learn-flink/fault_tolerance.zh.md [hunk 3]
```diff
@@ -29,180 +29,156 @@ under the License.
...
-### Exactly Once End-to-end
+### 端到端精确一次
 
-To achieve exactly once end-to-end, so that every event from the sources affects the sinks exactly
-once, the following must be true:
+为了实现端到端的精确一次，以便 sources 中的每个事件都仅精确一次对 sinks 生效，必须满足以下条件：
 
-1. your sources must be replayable, and
-2. your sinks must be transactional (or idempotent)
+1. 你的 sources 必须是可重播的，并且
```
Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:157]: `重播` -> `重放` 会好一些吗

File: docs/learn-flink/fault_tolerance.zh.md [hunk 4]
```diff
@@ -29,180 +29,156 @@ under the License.
 
 ## State Backends
 
-The keyed state managed by Flink is a sort of sharded, key/value store, and the working copy of each
-item of keyed state is kept somewhere local to the taskmanager responsible for that key. Operator
-state is also local to the machine(s) that need(s) it. Flink periodically takes persistent snapshots
-of all the state and copies these snapshots somewhere more durable, such as a distributed file
-system.
+由 Flink 管理的 keyed state 是一种分片的键/值存储，并且 keyed state 的每一项的工作副本都保存在负责该键的 taskmanager 本地的某个地方。
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply docs/learn-flink/fault_tolerance.zh.md:32]: @klion26 
`并且 keyed state 的每一项的工作副本`-->`每个 keyed state 的工作副本` ？

File: docs/learn-flink/fault_tolerance.zh.md [hunk 5]
```diff
@@ -29,180 +29,156 @@ under the License.
...
+Operator state 对于需要它的机器节点来说也是本地的。Flink 定期获取所有状态的连续快照，并将这些快照复制到持久化的地方，例如分布式文件系统。
 
-In the event of the failure, Flink can restore the complete state of your application and resume
-processing as though nothing had gone wrong.
+如果发生故障，Flink 可以恢复应用程序的完整状态并恢复处理，就如同没有出现过异常一样。
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply docs/learn-flink/fault_tolerance.zh.md:35]: @klion26 
`Flink 可以恢复应用程序的完整状态并恢复处理` --> `Flink可以恢复应用程序的完整状态并继续处理`?

**Comment by RocMarshal** [comment]
Hi, @klion26 ,
I have made some changes based on your suggestions, which is very helpful for improvement.
Thank you so much.

File: docs/learn-flink/fault_tolerance.zh.md [hunk 6]
```diff
@@ -29,180 +29,137 @@ under the License.
...
-In the event of the failure, Flink can restore the complete state of your application and resume
-processing as though nothing had gone wrong.
+如果发生故障，Flink 可以恢复应用程序的完整状态并继续处理，就如同没有出现过异常。
 
-This state that Flink manages is stored in a _state backend_. Two implementations of state backends
-are available -- one based on RocksDB, an embedded key/value store that keeps its working state on
-disk, and another heap-based state backend that keeps its working state in memory, on the Java heap.
-This heap-based state backend comes in two flavors: the FsStateBackend that persists its state
-snapshots to a distributed file system, and the MemoryStateBackend that uses the JobManager's heap.
+Flink 管理的状态存储 _state backend_。Flink 有两种 state backend 的实现 -- 一种基于 RocksDB 内嵌 key/value 存储将其工作状态保存在磁盘上的，另一种基于堆的 state backend，将其工作状态保存在 Java 的 堆内存中。这种基于堆的 state backend 有两种类型：FsStateBackend，将其状态快照持久化到分布式文件系统；MemoryStateBackend，它使用 JobManager 的堆保存状态快照。
```
Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:36]: `Flink 管理的状态存储 _state backend_` 这句话读起来会感觉不完整吗？

Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:36]:
```suggestion
Flink 管理的状态存储 _state backend_。Flink 有两种 state backend 的实现 -- 一种基于 RocksDB 内嵌 key/value 存储将其工作状态保存在磁盘上的，另一种基于堆的 state backend，将其工作状态保存在 Java 的堆内存中。这种基于堆的 state backend 有两种类型：FsStateBackend，将其状态快照持久化到分布式文件系统；MemoryStateBackend，它使用 JobManager 的堆保存状态快照。
```

File: docs/learn-flink/fault_tolerance.zh.md [hunk 7]
```diff
@@ -29,180 +29,137 @@ under the License.
...
-not past) both barriers.
+当 checkpoint coordinator（job manager 的一部分）指示 task manager 开始 checkpoint 时，它会让所有 sources 记录它们的偏移量，并将编号的 _checkpoint barriers_ 插入到它们的流中。这些屏障（barriers）流经 job graph，标注每个 checkpoint 前后的流部分。
 
-<img src="{{ site.baseurl }}/fig/stream_aligning.svg" alt="Barrier alignment" class="center" width="100%" />
+<img src="{% link fig/stream_barriers.svg %}" alt="Checkpoint barriers are inserted into the streams" class="center" width="80%" />
 
-Flink's state backends use a copy-on-write mechanism to allow stream processing to continue
-unimpeded while older versions of the state are being asynchronously snapshotted. Only when the
-snapshots have been durably persisted will these older versions of the state be garbage collected.
+Checkpoint _n_ 将包含每个 operator 的 state，这些 state 是对应的 operator 消费了 **严格在 checkpoint barrier _n_ 之前的所有事件，并且不包含在此（checkpoint barrier _n_）后的任何事件** 后而生成的状态。
```
Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:118]:
```suggestion
Checkpoint _n_ 将包含每个 operator 的 state，这些 state 是对应的 operator 消费了**严格在 checkpoint barrier _n_ 之前的所有事件，并且不包含在此（checkpoint barrier _n_）后的任何事件**后而生成的状态。
```

File: docs/learn-flink/fault_tolerance.zh.md (hunk 6 above)
Comment (RocMarshal) [reply docs/learn-flink/fault_tolerance.zh.md]: @klion26 
Thank you for your suggestions.
I changed it.

File: docs/learn-flink/fault_tolerance.zh.md [hunk 8]
```diff
@@ -29,180 +29,137 @@ under the License.
...
+- 没有任何丢失，但是你可能会得到重复冗余的结果（_at least once_）
+- 没有丢失或冗余重复（_exactly once_）
 
-To achieve exactly once end-to-end, so that every event from the sources affects the sinks exactly
-once, the following must be true:
+Flink 通过回退和重新发送 source 数据流从故障中恢复，当理想情况被描述为 **精确一次（exactly once）** 时，这并*不*意味着每个事件都将被精确一次处理。相反，这意味着 _每一个事件都会影响 Flink 管理的状态精确一次_。
 
-1. your sources must be replayable, and
-2. your sinks must be transactional (or idempotent)
+Barrier 只需要在提供精确一次的语义保证时需要对齐（Barrier alignment）。如果不需要这种语义，可以通过将 Flink 配置为使用 `CheckpointingMode.AT_LEAST_ONCE` 来获得一些性能，这具有关闭屏障对齐（Barrier alignment）的效果。
```
Comment (XBaith) [inline docs/learn-flink/fault_tolerance.zh.md:136]:
```suggestion
Barrier 只需要在提供精确一次的语义保证时需要对齐（Barrier alignment）。如果不需要这种语义，可以通过配置 `CheckpointingMode.AT_LEAST_ONCE` 关闭屏障对齐（Barrier alignment）来提高性能。
```

**Comment by RocMarshal** [comment]
Hi,@XBaith 
I have made some changes based on your suggestions, which is very helpful for improvement.
Thank you so much.

File: docs/learn-flink/fault_tolerance.zh.md [hunk 9]
```diff
@@ -29,180 +29,137 @@ under the License.
...
-into their streams. These barriers flow through the job graph, indicating the part of the stream
-before and after each checkpoint. 
+* _Savepoint_ -- 用户出于某种操作目的（例如有状态的重新部署/升级/缩放操作）手动（或 API 调用）触发的快照。Savepoints 始终是完整的，并且已针对操作灵活性进行了优化。
 
-<img src="{{ site.baseurl }}/fig/stream_barriers.svg" alt="Checkpoint barriers are inserted into the streams" class="center" width="80%" />
+### 状态快照如何工作？
 
-Checkpoint _n_ will contain the state of each operator that resulted from having consumed **every
-event before checkpoint barrier _n_, and none of the events after it**.
+Flink 使用 [Chandy-Lamport algorithm](https://en.wikipedia.org/wiki/Chandy-Lamport_algorithm) 算法的一种变体，称为异步屏障快照（_asynchronous barrier snapshotting_）。
```
[the remark this answers is not in the document]
Comment (klion26) [reply docs/learn-flink/fault_tolerance.zh.md:112]: barrier 个人觉得不翻译也可以

File: docs/learn-flink/fault_tolerance.zh.md [hunk 10]
```diff
@@ -29,180 +29,137 @@ under the License.
...
-Flink exactly once_. 
+### 确保精确一次（exactly once）
 
-Barrier alignment is only needed for providing exactly once guarantees. If you don't need this, you
-can gain some performance by configuring Flink to use `CheckpointingMode.AT_LEAST_ONCE`, which has
-the effect of disabling barrier alignment.
+当流处理应用程序发生错误的时候，结果可能会产生丢失或者重复。Flink 根据你为应用程序和集群的配置，可以产生以下结果：
 
-### Exactly Once End-to-end
+- Flink 不会努力从故障中恢复（_at most once_）
```
Comment (klion26) [inline docs/learn-flink/fault_tolerance.zh.md:130]: 这个地方再次看的时候，可以讨论下这个翻译是否需要优化下。“不会努力从故障中恢复” 是否有一种“不会恢复” 的意思呢？这里我们应该要着重体现“不从 checkpoint 中恢复” 这个事情。这里能否直接翻译成 “不会从快照中进行恢复”呢？

**Comment by RocMarshal** [comment]
@klion26 , @XBaith 
Thank you for your help.
I have made some alterations according to your suggestions, please take a look.

**Comment by RocMarshal** [comment]
@klion26 , @XBaith 
Thank you for your help.
I have made some alterations according to your suggestions, please take a look.

File: docs/learn-flink/fault_tolerance.zh.md (hunk 10 above)
Comment (XBaith) [reply docs/learn-flink/fault_tolerance.zh.md]: 是的你是对的，我之前建议的应该不对

**Comment by XBaith** [comment]
Thank you @RocMarshal. LGTM.

File: docs/learn-flink/fault_tolerance.zh.md [hunk 11]
```diff
@@ -29,180 +29,156 @@ under the License.
...
-checkpoint, it has all of the sources record their offsets and insert numbered _checkpoint barriers_
-into their streams. These barriers flow through the job graph, indicating the part of the stream
-before and after each checkpoint. 
+### 状态快照如何工作？
+
+Flink 使用 [Chandy-Lamport algorithm](https://en.wikipedia.org/wiki/Chandy-Lamport_algorithm) 算法的一种变体，称为异步屏障快照（_asynchronous barrier snapshotting_）。
+
+当 checkpoint coordinator（job manager 的一部分）指示 task manager 开始 checkpoint 时，它会让所有 sources 记录它们的偏移量，并将编号的 _checkpoint barriers_ 插入到它们的流中。这些屏障（barriers）流经作业图（job graph），标注每个 checkpoint 前后的流部分。
 
 <img src="{{ site.baseurl }}/fig/stream_barriers.svg" alt="Checkpoint barriers are inserted into the streams" class="center" width="80%" />
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply docs/learn-flink/fault_tolerance.zh.md]: @klion26 , @wuchong .
Should **"{{ site.baseurl }}/fig/stream_barriers.svg"** be written as **"{% link fig/stream_barriers.svg %}"** or **"/fig/stream_barriers.svg"** according to **Visual Elements** in [docs-style](https://flink.apache.org/zh/contributing/docs-style.html) ?

**Comment by wuchong** [comment]
Thanks for the great work @RocMarshal @klion26  @XBaith. 

Merging...

File: docs/learn-flink/fault_tolerance.zh.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (wuchong) [reply docs/learn-flink/fault_tolerance.zh.md]: It is recommended to use `{% link fig/stream_barriers.svg %}` which will the check the link validity. This is discussed recently in dev ML, but not update the doc guideline yet.

[the remark this answers is not in the document]
Comment (RocMarshal) [reply docs/learn-flink/fault_tolerance.zh.md]: @wuchong ,OK.
Thank you for your reply.

**Pull request closed without merging**